pkg/resourcevalidator/sp_description.go

```diff
@@ -0,0 +1,105 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resourcevalidator
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	vkv1alpha1 "github.com/liqotech/liqo/apis/virtualkubelet/v1alpha1"
+)
+
+// ShadowPodDescription is a struct that contains the main informations about a shadow pod.
+type ShadowPodDescription struct {
+	NamespacedName types.NamespacedName
+	UID            types.UID
+	Quota          v1.ResourceList
+	timestamp      string
+	running        bool
+}
+
+func createShadowPodDescription(name, namespace string, uid types.UID, resources v1.ResourceList) ShadowPodDescription {
+	return ShadowPodDescription{
+		NamespacedName: types.NamespacedName{Name: name, Namespace: namespace},
+		UID:            uid,
+		Quota:          resources.DeepCopy(),
+		timestamp:      time.Now().Format(time.RFC3339),
+		running:        true,
+	}
+}
+
+func (spd *ShadowPodDescription) terminate() {
+	spd.running = false
+	spd.timestamp = time.Now().Format(time.RFC3339)
+}
+
+func getOrCreateShadowPodDescription(ctx context.Context, c client.Client, pi *peeringInfo, sp *vkv1alpha1.ShadowPod) (*ShadowPodDescription, error) {
+	pi.piMutex.Lock()
+	defer pi.piMutex.Unlock()
+	key, _ := cache.MetaNamespaceKeyFunc(types.NamespacedName{Name: sp.Name, Namespace: sp.Namespace})
+	spd, found := pi.getShadowPodDescription(key)
+	shadowpodlog.Info(fmt.Sprintf("ShadowPod: %v", found))
+	spQuota, err := getQuotaFromShadowPod(sp, true)
+	if err != nil {
+		return nil, err
+	}
+	shadowpodlog.Info(quotaFormatter(spQuota, "ShadowPod quota"))
+	if !found {
+		spd = createShadowPodDescription(sp.GetName(), sp.GetNamespace(), sp.GetUID(), spQuota)
+	} else {
+		if spd.running {
+			return nil, fmt.Errorf("ShadowPod %s is already running", sp.GetName())
+		}
+		found, err := checkShadowPodExistence(ctx, c,
+			types.NamespacedName{Name: sp.GetName(), Namespace: sp.GetNamespace()})
+		if found {
+			return nil, fmt.Errorf("ShadowPod is terminating but still exists in the system")
+		} else if err != nil {
+			return nil, err
+		}
+		pi.removeShadowPod(spd)
+		spd = createShadowPodDescription(sp.GetName(), sp.GetNamespace(), sp.GetUID(), spQuota)
+	}
+	return &spd, nil
+}
+
+func getOrDeleteShadowPodDescription(pi *peeringInfo, sp *vkv1alpha1.ShadowPod) (*ShadowPodDescription, error) {
+	pi.piMutex.Lock()
+	defer pi.piMutex.Unlock()
+	key, _ := cache.MetaNamespaceKeyFunc(types.NamespacedName{Name: sp.Name, Namespace: sp.Namespace})
+	spd, found := pi.getShadowPodDescription(key)
+	check := spd.UID == sp.GetUID()
+	if !found {
+		spQuota, _ := getQuotaFromShadowPod(sp, false)
+		// Create a new ShadowPodDescription for caching purposes
+		spd = createShadowPodDescription(sp.GetName(), sp.GetNamespace(), sp.GetUID(), spQuota)
+		spd.terminate()
+		pi.updateShadowPod(spd)
```

Comment:
No because in this case there is a mismatch between the request and the cache status, so it update the cache in order to allow the "refresher" to align it successively without touching the available/used resources that are already aligned.